### ISSUE TITLE
Learning Rate Scheduler

### ISSUE THREAD

**Issue by jafioti** [coreylowman/dfdx#439]
It would be really useful to have a learning rate scheduler trait to implement various schedules.

The trait could be as general as something like:
```rust
trait OptimScheduler<T> {
    fn step(&mut self, optimizer: &mut T);
}
```
where each time step is called, the optimizer is passed in and changed according to the scheduler.

**Comment by coreylowman** [comment]
Thoughts on this being implemented on the optimizer itself rather than a separate object? It could help alleviate all the issues pytorch had with when to call lr_scheduler.step()

**Comment by jafioti** [comment]
> Thoughts on this being implemented on the optimizer itself rather than a separate object? It could help alleviate all the issues pytorch had with when to call lr_scheduler.step()

The issue with having it on the same object is that the optimizer and the schedule may be completely different. For instance, the Adam optimizer can be used with many different schedules, and it would be great if people can develop their schedulers separate from the library and optimizer. There are new schedulers coming out every day in papers, so it doesn't make sense to tie an optimizer to a single scheduler.

**Comment by nkoppel** [comment]
> There are new schedulers coming out every day in papers, so it doesn't make sense to tie an optimizer to a single scheduler.

It should be possible to generalize learning rate schedules with a trait, and then have those as a field of the optimizer. For example:

```rust
trait LearningRateSchedule<E: Dtype> {
    fn next(&mut self) -> E;
}

impl LearningRateSchedule<f32> for f32 {
    fn next(&mut self) -> f32 {
        *self
    }
}

struct LinearLrSchedule<E: Dtype> {
    value: E,
    slope: E
}

impl LearningRateSchedule<E> for LinearLrSchedule<E> {
    fn next(&mut self) {
        self.value += self.slope;
        self.value
    }
}
```
This would also not be a breaking change to the api, since f32 would act the same as a lr as a lr schedule. The only potential issue with this is that the user may not want to update the learning rate schedule for each optimizer step.

**Comment by jafioti** [comment]
@nkoppel I think the issue with that is A) the learning rate scheduler must be ran each iteration, and B) you can't build schedulers based on external factors, such as a scheduler that decreases the LR if no loss decrease happened over the past 100 iters.

**Comment by jafioti** [comment]
I wrote a OneCycle scheduler and a LearningRate trait that can be implemented on any scheduler with a learning rate:
```rust
pub trait LearningRate<E> {
    fn learning_rate(&mut self) -> &mut E;
}

impl<M, E: Dtype, D: DeviceStorage> LearningRate<E> for Adam<M, E, D> {
    fn learning_rate(&mut self) -> &mut E {
        &mut self.cfg.lr
    }
}

pub trait OptimScheduler<T> {
    fn step(&mut self, optimizer: &mut T);
}

pub struct OneCycleScheduler<E> {
    base_lr: E,
    max_lr: E,
    progress: f32,
}

impl<E> OneCycleScheduler<E> {
    pub fn new(base_lr: E, max_lr: E) -> Self {
        Self {
            base_lr,
            max_lr,
            progress: 0.,
        }
    }

    pub fn set_progress(&mut self, progress: f32) {
        self.progress = progress;
    }
}

impl<E: Dtype, L: LearningRate<E>> OptimScheduler<L> for OneCycleScheduler<E> {
    fn step(&mut self, optimizer: &mut L) {
        let progress = (0.5 - (self.progress - 0.5).abs()) * 2.;
        *optimizer.learning_rate() = 
            (self.max_lr - self.base_lr)
             * E::from_f32(progress).unwrap()
             + self.base_lr;
    }
}
```
I think as a demo of the OptimScheduler trait it seems quite simple and flexible. Much more complex schedulers can be implemented with this trait. The only thing I don't really like too much is the LearningRate trait, because we'll need that and other traits to expose certian variables in different optimizers. LearningRate is simple enough, but perhaps there will be schedulers that change momentum as well, so we'll need a Momentum trait.

Here's the API to use it:
```rust
fn train_epoch<D: Device<f32>, O>(
    model: &mut Model, 
    opt: &mut O,
    scheduler: &mut OneCycleScheduler<f32>,
    dev: &D,
) where 
    D: Device<f32>,
    O: Optimizer<Model, D, f32> + LearningRate<f32>,
    [(); EMBED * 2]:
{
    // Training stuff...
    scheduler.set_progress(percent_of_epoch_complete);
    scheduler.step(opt);
    opt.update(model, &grads).unwrap();
}
```
